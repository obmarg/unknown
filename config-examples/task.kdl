// Thinking this would allow > 1 task per task file
// Which is probably good?  I enjoy it in makefiles since
// tasks generally aren't overly complicated.
// Down with too many files.

task "xyz" {
    command "cargo build"
}