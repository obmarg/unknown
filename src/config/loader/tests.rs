use miette::{GraphicalReportHandler, GraphicalTheme};

use crate::test_files::TestFiles;

use super::*;

#[test]
fn test_load_config_from_cwd() {
    insta::assert_debug_snapshot!(load_config_from_path("sample-monorepo/".into()));
}

#[test]
fn test_malformed_project_file() {
    let test_files = TestFiles::new()
        .with_file(
            "workspace.kdl",
            r#"
                name "my-workspace"
                project_path "**"
            "#,
        )
        .with_file(
            "project/project.kdl",
            r#"
                tasks {
                    import "missing.nabs"
                }
            "#,
        );

    let err = load_config_from_path(test_files.root().into()).unwrap_err();

    insta::assert_display_snapshot!(render_miette(err, &test_files));
}

#[test]
fn test_missing_task_file_error() {
    let test_files = TestFiles::new()
        .with_file(
            "workspace.kdl",
            r#"
                name "my-workspace"
                project_path "**"
            "#,
        )
        .with_file(
            "project/project.kdl",
            r#"
            project "missing-task-file-project"

            tasks {
                import "missing.nabs"
            }
        "#,
        );

    let err = load_config_from_path(test_files.root().into()).unwrap_err();

    insta::assert_display_snapshot!(render_miette(err, &test_files));
}

#[test]
fn test_dependency_from_out_of_workspace() {
    let test_files = TestFiles::new()
        .with_file("project.kdl", r#"project "proj-outside-workspace"#)
        .with_file(
            "workspace/workspace.kdl",
            r#"
                name "test"

                project_path "*"
            "#,
        )
        .with_file(
            "workspace/project/project.kdl",
            r#"
                project "my_project"
                dependencies {
                    project "../../"
                }
            "#,
        );

    let err =
        load_config_from_path(Utf8PathBuf::from(test_files.root()).join("workspace")).unwrap_err();

    insta::assert_display_snapshot!(render_miette(err, &test_files));
}

#[test]
fn test_wont_import_tasks_from_out_of_workspace() {
    let test_files = TestFiles::new()
        .with_file("task.kdl", "")
        .with_file(
            "workspace/workspace.kdl",
            r#"
                name "test"

                project_path "*"
            "#,
        )
        .with_file(
            "workspace/project/project.kdl",
            r#"
                project "my_project"
                tasks {
                    import "../../task.kdl"
                }
            "#,
        );
    let err =
        load_config_from_path(Utf8PathBuf::from(test_files.root()).join("workspace")).unwrap_err();

    insta::assert_display_snapshot!(render_miette(err, &test_files));
}

fn render_miette(e: miette::Report, test_files: &TestFiles) -> String {
    let mut report = String::new();
    GraphicalReportHandler::new_themed(GraphicalTheme::unicode_nocolor())
        .render_report(&mut report, e.as_ref())
        .unwrap();

    // We don't want any randomly generated absolute paths in our error, so
    // find the root and replace it in the error string with /
    let file_root = Utf8PathBuf::from(test_files.root())
        .canonicalize_utf8()
        .unwrap();

    report.replace(file_root.as_str(), "/")
}
