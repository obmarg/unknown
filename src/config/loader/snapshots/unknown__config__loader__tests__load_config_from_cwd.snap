---
source: src/config/loader/tests.rs
expression: "load_config_from_path(\"sample-monorepo/\".into()).unwrap()"
---
ValidConfig {
    workspace_file: WorkspaceFile {
        workspace_root: WorkspaceRoot(
            "/Users/graeme/src/unknown/sample-monorepo/",
        ),
        config: WorkspaceDefinition {
            name: "my-workspace",
            project_paths: [
                Glob(
                    Glob {
                        glob: "**",
                        re: "(?-u)^.*$",
                        opts: GlobOptions {
                            case_insensitive: false,
                            literal_separator: false,
                            backslash_escape: true,
                        },
                        tokens: Tokens(
                            [
                                RecursivePrefix,
                            ],
                        ),
                    },
                ),
            ],
        },
        source: ConfigSource {
            filename: "workspace.kdl",
            ..
        },
    },
    project_files: [
        ValidProjectFile {
            project_root: ValidPath {
                workspace_root: WorkspaceRoot(
                    "/Users/graeme/src/unknown/sample-monorepo/",
                ),
                subpath: "projects/a-service",
            },
            config: ProjectDefinition {
                project: "a-service",
                dependencies: [
                    ValidPath {
                        workspace_root: WorkspaceRoot(
                            "/Users/graeme/src/unknown/sample-monorepo/",
                        ),
                        subpath: "projects/a-lib",
                    },
                ],
                tasks: TaskBlock {
                    imports: [],
                    tasks: [
                        TaskDefinition {
                            name: "build",
                            commands: [
                                "echo \"build-a-service\"",
                            ],
                            requires: [
                                TaskRequires {
                                    task: "build",
                                    target: Some(
                                        DependenciesOfCurrent,
                                    ),
                                },
                            ],
                            input_blocks: [],
                            source: ConfigSource {
                                filename: "projects/a-service/project.kdl",
                                ..
                            },
                        },
                        TaskDefinition {
                            name: "bye",
                            commands: [
                                "echo \"bye\"",
                            ],
                            requires: [],
                            input_blocks: [],
                            source: ConfigSource {
                                filename: "projects/a-service/bye.nabs",
                                ..
                            },
                        },
                    ],
                },
            },
            source: ConfigSource {
                filename: "projects/a-service/project.kdl",
                ..
            },
        },
        ValidProjectFile {
            project_root: ValidPath {
                workspace_root: WorkspaceRoot(
                    "/Users/graeme/src/unknown/sample-monorepo/",
                ),
                subpath: "projects/a-lib",
            },
            config: ProjectDefinition {
                project: "a-lib",
                dependencies: [],
                tasks: TaskBlock {
                    imports: [],
                    tasks: [
                        TaskDefinition {
                            name: "build",
                            commands: [
                                "echo \"build-a-lib\"",
                            ],
                            requires: [],
                            input_blocks: [],
                            source: ConfigSource {
                                filename: "projects/a-lib/project.kdl",
                                ..
                            },
                        },
                        TaskDefinition {
                            name: "hello",
                            commands: [
                                "echo \"hello\"",
                            ],
                            requires: [],
                            input_blocks: [],
                            source: ConfigSource {
                                filename: "tasks/hello.nabs",
                                ..
                            },
                        },
                        TaskDefinition {
                            name: "hello2",
                            commands: [
                                "echo \"hello\"",
                            ],
                            requires: [],
                            input_blocks: [],
                            source: ConfigSource {
                                filename: "tasks/hello2.nabs",
                                ..
                            },
                        },
                    ],
                },
            },
            source: ConfigSource {
                filename: "projects/a-lib/project.kdl",
                ..
            },
        },
    ],
}
